### PR TITLE
Route Callbacks/Filters

### PR DESCRIPTION
**Pull request by shadowhand**
http://dev.kohanaframework.org/issues/4388

### CODE REVIEW

**Comment by shadowhand** [comment]
This probably needs a test or two before merging.

**Comment by kiall** [comment]
Dont expect a comment from Jenkins anytime soon! will be offline for a bit.. HW node panicked 3 times last night due to memory issues.. trying to identify which stick is bad now.. memtest will likely take a few hours ;)

**Comment by kiall** [comment]
+1 on this change btw, looks like it could be quite useful for APIs, unless we give routes a more direct way to handle HTTP methods..

Also, I think @Zeelot has something similar in hyla. Zeelot: how does this compare?

**Comment by shadowhand** [comment]
I think @Zeelot is running the callbacks from inside `Request` so his callback signature looks like `function(Request $request, Route $route, $params)`... but he can confirm that.

I still want a different way of handling HTTP methods.

**Comment by shadowhand** [comment]
@Zeelot's way is probably the correct solution, as it the most flexible without modifying `Route::matches()`.

**Comment by shadowhand** [comment]
The [code in question](https://github.com/hyla/hyla/blob/1.0.x/develop/hyla/core/classes/route.php). In this code, Request has been modified to send itself to `Route::matches()` instead of the URI. I think that might be too large of change to make at this point, unless it fits in with #194 somehow.

**Comment by zeelot** [comment]
Yeah I prefer having access to the entire request object so that the callbacks can look at more than just the URI which I think is important.

**Comment by zeelot** [comment]
I am allowing the user to specify the parameters for the callbacks similarly to Validation rules but I'm not sure if you guys find that useful or if we should just always pass in the same things. The reason I had it was so you could reuse callbacks a little more easily like this:

```
->callback(array('Router', 'match_methods'), array(':request', array('GET', 'POST')))
```

I know you want a different way of handling http methods but maybe it's still useful for other things.

**Comment by zeelot** [comment]
Looks like you're currently forcing the user to return the array of params on all filters, what do you think of just replacing params if the callback returns an array? From my experience, there are 2 types of callbacks I run; callbacks to enforce a specific thing (return FALSE in a condition, do nothing otherwise), and callbacks that slightly alter the params (if id is not set, change the action)

**Comment by zeelot** [comment]
Sorry last comment =P I don't think the changes to Route and Request I made in Hyla are too large, In my opinion, routes need to match an entire request instead of just the URI so I would really like the change to make it in. I'm not sure at this point how much work it is with all the changes that @kiall has made (stopped keeping track to the request work)

**Comment by kiall** [comment]
I'm going to try and break that up into a few smaller pulls ;)

**Comment by shadowhand** [comment]
@Zeelot wrote:

> what do you think of just replacing params if the callback returns an array?

That is probably the least error prone way to do it, good idea.

> In my opinion, routes need to match an entire request instead of just the URI so I would really like the change to make it in.

I agree with this as well, it makes a lot of sense to do `Route::matches(Request $request)`.

At this point, I am not sure what to do about the variable parameters, it seems like a lot of extra overhead for something that could be easily accomplished by creating a closure that calls another method.

@kiall The build failed on something completely unrelated... any idea what that is about?

**Comment by kiall** [comment]
Yea - Looks like I managed to move the wrong version of the HTTP_Cache class from core -> cache.. I think I grabbed the copy from my request/resonse/HVMC isolation stuff. Sorting it now..

**Comment by kiall** [comment]
!build

**Comment by shadowhand** [comment]
If everyone is in agreement, I vote for this to be merged right now, then we can start looking at changing how `Route::matches()` is called.

**Comment by shadowhand** [comment]
!build

**Comment by kiall** [comment]
Just a FYI, "build" is only needed to force a re-test when there have been no code changes, if you add/edit commits it will pick it up on the next check (every 15 mins)

**Comment by shadowhand** [comment]
@kiall Okay, good to know.

File: classes/kohana/route.php [hunk 1]
```diff
@@ -359,6 +364,42 @@ public function defaults(array $defaults = NULL)
...
+	 * [!!] Default parameters are added before filters are called!
+	 *
+	 * @throws  Kohana_Exception
+	 * @param   array   $callback   callback string, array, or closure
+	 * @return  $this
+	 */
+	public function filter($callback)
+	{
+		if ( ! is_callable($callback))
+		{
```
Comment (kiall) [inline classes/kohana/route.php:393]: Minor Coding Style issue, the brackets aren't required for return/throw etc.. From memory I think our rules pick up on this..

Comment (shadowhand) [inline classes/kohana/route.php:393]: Ah crap, I already merged. Can be fixed in develop branch if CS picks it up.

**Pull request merged**